### PR TITLE
blob: add a separate Attributes API

### LINKED ISSUE #455: blob: separate out stat/get attributes API
Today you can access attributes by creating a Reader, then calling methods on the Reader to get attributes. To only get attributes, you do a 0-length read. This is kind of a weird API, and for both S3 and GCS, it results in a different HTTP request for a 0-length read. Even worse, the GCS API returns a different set of attributes for the 0-length read than for the full read (example: we can't get the Metadata via the full read, see #324).

Proposal:

- Return a small subset of attributes, probably Size and ContentType, from Reader.
- Create a new concrete type method (and possibly Driver method to match) that explicitly returns attributes.

### PR DESCRIPTION
**Pull request by vangent**
Fixes #455.

### CODE REVIEW

File: blob/blob.go [hunk 1]
```diff
@@ -176,6 +181,9 @@ func (b *Bucket) NewRangeReader(ctx context.Context, key string, offset, length
 	if offset < 0 {
 		return nil, errors.New("new blob range reader: offset must be non-negative")
 	}
+	if length == 0 {
+		return nil, errors.New("new blob range reader: length cannot be 0")
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/blob.go]: I don't think it makes sense to do a 0-length read, and provider behavior may vary when asked to do so. I'd rather just fail in the concrete type like this, to be consistent.

I updated the comments to note this, and to tell the user that they should use Attributes if they want to check for existence.

File: blob/gcsblob/gcsblob.go [hunk 2]
```diff
@@ -71,52 +69,45 @@ func (r *reader) Close() error {
...
-// the metadata. If length is negative, it will read till the end of the object.
-func (b *bucket) NewRangeReader(ctx context.Context, key string, offset, length int64) (driver.Reader, error) {
-	if offset < 0 {
-		return nil, fmt.Errorf("negative offset %d", offset)
-	}
+func (r *reader) Size() int64 {
+	return r.size
+}
+
+func (b *bucket) Attributes(ctx context.Context, key string) (driver.Attributes, error) {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/gcsblob/gcsblob.go]: Updated all of the driver implementations to just reference the corresponding driver.Driver method. There's no need to repeat the docstring, the interface should have the full description.

File: blob/gcsblob/gcsblob.go [hunk 3]
```diff
@@ -71,52 +69,45 @@ func (r *reader) Close() error {
...
+		return driver.Attributes{}, err
 	}
+	return driver.Attributes{
+		ContentType: attrs.ContentType,
+		ModTime:     attrs.Updated,
+		Size:        attrs.Size,
+	}, nil
+}
+
+func (b *bucket) NewRangeReader(ctx context.Context, key string, offset, length int64) (driver.Reader, error) {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/gcsblob/gcsblob.go]: Done.

File: blob/driver/driver.go [hunk 4]
```diff
@@ -42,10 +42,10 @@ type Error interface {
 type Reader interface {
 	io.ReadCloser
 
-	// Attrs returns the object's metadata. It may return a different pointer each
-	// time, but it must return the exact same values each time it is called. The
-	// caller must not modify any fields in the returned ObjectAttrs.
-	Attrs() *ObjectAttrs
+	// ContentType is the MIME type of the blob object. It must not be empty.
+	ContentType() string
```
Comment (zombiezen) [inline blob/driver/driver.go]: Making these be methods makes introducing a field be more complicated to roll out. I suggest having two different structs: one that is the "reader attributes" and one that is "all attributes". The latter may embed the former.

Comment (vangent) [reply blob/driver/driver.go]: Great point, thanks. Done.

File: blob/blob.go [hunk 5]
```diff
@@ -44,21 +43,19 @@ func (r *Reader) Close() error {
 	return r.r.Close()
 }
 
-// ContentType returns the MIME type of the blob object.
-func (r *Reader) ContentType() string {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/blob.go]: Done.

File: blob/blob.go [hunk 6]
```diff
@@ -44,21 +43,19 @@ func (r *Reader) Close() error {
...
 }
 
-// Size returns the content size of the blob object.
-func (r *Reader) Size() int64 {
-	return r.r.Attrs().Size
-}
+// ReaderAttributes contains a subset of attributes about a blob that are
+// accessible from Reader. Use Bucket.Attributes to get the full set of
+// attributes.
+type ReaderAttributes driver.ReaderAttributes
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/blob.go]: ReaderAttributes is no longer in the concrete type.

I made Attributes a type alias.

File: blob/blob.go [hunk 7]
```diff
@@ -166,16 +172,21 @@ func (b *Bucket) NewReader(ctx context.Context, key string) (*Reader, error) {
 }
 
 // NewRangeReader returns a Reader that reads part of an object, reading at
-// most length bytes starting at the given offset. If length is 0, it will read
-// only the metadata. If length is negative, it will read till the end of the
-// object. It returns an error if that object does not exist, which can be
+// most length bytes starting at the given offset. If length is negative, it
+// will read till the end of the object.
+// offset must be >= 0, and length cannot be 0; use Attributes to check for
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/blob.go]: Done.

File: blob/drivertest/drivertest.go [hunk 8]
```diff
@@ -86,8 +85,8 @@ func testRead(t *testing.T, newHarness HarnessMaker) {
 			wantErr: true,
 		},
 		{
-			name:    "length 0 read of nonexistent key fails",
-			key:     "key-does-not-exist",
+			name:    "length 0 read fails",
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/drivertest/drivertest.go]: Yes. I didn't think it was worth checking the nonexistent case separately.

In practice, this check is done in the concrete type, so it won't vary from provider to provider.

File: blob/driver/driver.go [hunk 9]
```diff
@@ -61,24 +60,40 @@ type WriterOptions struct {
...
+
+// Attributes contains attributes about a blob.
+type Attributes struct {
 	// ContentType is the MIME type of the blob object. It must not be empty.
 	ContentType string
+	// Size is the size of the object in bytes.
+	Size int64
+
 	// ModTime is the modified time of the blob object. Will be time.Time zero value if unknown.
 	ModTime time.Time
```
Comment (zombiezen) [inline blob/driver/driver.go]: Wasn't this already a reader attribute? Why the change? (Apologies if this was already addressed elsewhere.)

Comment (vangent) [reply blob/driver/driver.go]: Yes, it was a Reader attribute, and I could put it in ReaderAttributes. I didn't largely because GCS didn't originally support it; Tuo filed a feature request and they added it for us. I take this as a signal that it's an attribute that not all providers might support as part of reading.

I don't feel strongly about this either way though.

Comment (zombiezen) [reply blob/driver/driver.go]: I feel like it should be included, because otherwise we're discarding the information. We already document that it might not be filled in.

Comment (vangent) [reply blob/driver/driver.go]: Done.
